File: dynamic-programming/rod-cutting/golang/recursive_rod_cutting.go
package main

import (
	"fmt"
	"math"
)

func cutRod(prices []int, length int) int {
	if length == 0 {
		return 0
	}

	result := -math.MaxInt64
	for i := 1; i < length+1; i++ {
		result = max(result, prices[i]+cutRod(prices, length-i))
	}

	return result
}

func main() {
	prices := []int{0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30}

	fmt.Print("  i |")
	for i := range prices {
		fmt.Printf("%3d|", i)
	}
	fmt.Print("\n")
	fmt.Print("p[i]|")
	for _, p := range prices {
		fmt.Printf("%3d|", p)
	}
	fmt.Print("\n")

	for i := range prices {
		result := cutRod(prices, i)

		fmt.Printf("length = %d / result = %d\n", i, result)
	}
}

File: dynamic-programming/rod-cutting/python/recursive_rod_cutting.py
import sys


def cut_rod(prices, length):
    if length == 0:
        return 0

    result = -sys.maxsize
    for i in range(1, length + 1):
        result = max(result, prices[i] + cut_rod(prices, length - i))

    return result


if __name__ == "__main__":
    prices = [0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30]

    print("  i |", end="")
    for i in range(len(prices)):
        print(f"{i:3}|", end="")
    print("")
    print("p[i]|", end="")
    for p in prices:
        print(f"{p:3}|", end="")
    print("")

    for n in range(len(prices)):
        result = cut_rod(prices, n)

        print(f"length = {n} / result = {result}")
